interview_questions.md

```diff
@@ -11,127 +11,916 @@ and Rails, grouped by category.
+  ```
+  To search a gem :-
+  gem search -r KEYWORD
+
+  To install a gem :-
+  gem install GEM_NAME
+  ```
+
 - How to define functions?
+  ```
+  Without Variables :-
+  def FUNCTION_NAME
+    # Code
+  end
+
+  With Variables :-
+  def FUNCTION_NAME(VARIABLE_NAME=DEFAULT_VALUE)
+    # Code
+  end
+  ```
+
 - What are predicate functions?
+  ```
+  A method that returns true or false is called a “predicate method”. 
+  In Ruby, there is a naming convention where predicate methods end with a question mark.
+  ```
+
 - What are safe and unsafe functions?
+  ```
+  The functions ending with ! (save!, update! and others) are called unsafe functions and raise an exception if the record is not valid.
+
+  On the other hand, functions not ending with ! are called safe functions and return true/false depending whether the record is valid.
+  ```
+
 - What are instance and class variables?
+  ```
+  Class variables are shared between a class and all its subclasses, while Instance variables only belong to one specific class.
+  ```
+
 - What are instance and class functions?
+  ```
+  A class method provides functionality to a class itself, while an instance method provides functionality to one instance of a class.
+  We cannot call an instance method on the class itself, and we cannot directly call a class method on an instance.
+  ```
+
 - What are attribute readers, writers and accessors?
+  ```
+  'attr_accessor' allows instance variables to be accessed outside the methods.
+
+  'attr_accessor' creates both the reader & writer methods.
+  'attr_reader' creates only the reader.
+  'attr_writer' creates only the writer.
+  ```
+
 - What are strings and symbols?
+  ```
+  Strings are any text written between quotation marks ("string”), while a symbol is text that begins with a colon (:symbol)
+
+  We use Symbols when the value will remain constant. For example, hash keys, while we use Strings when the value or you want to change the value or use a string method on it
+  ```
+
 - What is a class? How to define a class?
+  ```
+  A Class is a blueprint from which individual distinct objects can be created.
+
+  To define a class in Ruby :-
+  class ClassName
+  end
+
+  To create a object of this class :-
+  object = ClassName.new
+  ```
+
 - What are some different ways to iterate over an array?
+  ```
+  Different ways to iterate over an array :-
+  1. While Loop
+    while EXPRESSION
+      # Some Code
+    end
+
+  2. For Loop
+    for element in Array_name
+      # Some Code
+    end
+
+  3. Each Loop
+    Array_name.each |element|
+      # Some Code
+    end
+  ```
+
 - How to create a hash variable and iterate over it?
+  ```
+  Creating a hash variable :-
+  hash = {
+    key1 => value1,
+    key2 => value2
+  }
+
+  Iterating over hash :-
+  hash.each do |key, value|
+    # Some Code
+  end
+  ```
+
 - What are some differences between `nil` and `false`?
+  ```
+  nil means no data value but memory is still allocated for null. 
```

Comment:
This is not the database value `null` but the ruby object `nil` - incorrect answer and refer to this: https://learn.co/lessons/truthiness-in-ruby-readme and https://gist.github.com/jfarmer/2647362

Reply:
Ig it got auto-correct from nil to null 😄..ill correct that